### ISSUE TITLE
No matching handler found for my mock

### ISSUE THREAD

**Issue by robertmain** [mockery/mockery#856]
Currently trying to mock a method in my parent class out. I set up the expectation but I get an error from Mockery saying:

``
Mockery\Exception\NoMatchingExpectationException: No matching handler found for Mockery_1_App_Models_User::update(6, ['password' => '$2y$10$3zhUkgt9OlneZAhgku62TuDIKgT1XQFfiaUeb4HTrI3xHuoHybCai', 'updated_at' => '2018-02-26 00:53:43']). Either the method was unexpected or its arguments matched no expected argument list for this method
``

Somehow, when I'm calling `User::save()` with the password, it's calling the parent save method which is then trying to call the `update` method - but somehow my expecting isn't being hit.

My tests for `User::password_verify` pass which depends on the parent model's `get_by` method being mocked out. I can't see what's different here and why this doesn't work..

Am I missing something?

Model.php:

```php
<?php

namespace App\Core;

use CI_Model;

/**
 * Abstract Base Model
 *
 * Abstract database model. Adds additional functionality such as created/updated/deleted timestamps, etc.
 */
abstract class Model extends CI_Model
{
    /**
     * @var string Table primary key column
     */
    const PRIMARY_KEY = 'id';

    /**
     * @var string The name for the field used to store record creation timestamp
     */
    const CREATED = 'created_at';

    /**
     * @var string The name for the field used to store record update timestamp
    */
    const UPDATED = 'updated_at';

    /**
     * @var string The name of the table data for this model is stored in. A protected variable is used to store this
     *             information instead of a class constant, as this value is computed at call time
     */
    protected $table = null;

    /**
     * Dynamically set the model's database table name (though, this can be overridden..)
     */
    public function __construct()
    {
        parent::__construct();

        $this->load->helper(['inflector']);

        $this->load->database();

        if ($this->table === null) {
            $className = (new \ReflectionClass($this))->getShortName();
            $this->table = strtolower(plural($className));
        }
    }

    /**
     * Update a single record in the database
     *
     * @param mixed $primary_value The primary key value of the record to update
     * @param array $data          An ascociative array of update values
     *
     * @return boolean Indication of the success of the update
     */
    protected function update($primary_value, $data) : \boolean
    {
        $data = $this->run_before_callbacks('update', [$data, $primary_value]);

        $result = $this->db->where(static::PRIMARY_KEY, $primary_value)
                           ->update($this->table, $data);

        $this->run_after_callbacks('update', [$data, $primary_value, $result]);

        return $result;
    }

    /**
     * Saves (or creates) a single record if it doesn't already exist. If the `$id1 parameter is provided, the record
     * is presumed not to exist and is created. The newly created ID is then returned.
     *
     * @param array $data The data to persist to the database
     * @param int   $id   The primary key value of the record to edit (if omitted, create is assumed)
     *
     * @return int The primary key of the record that was updated/deleted
    */
    public function save($data, $id = null)
    {
        $date = new \DateTime();

        $data[static::UPDATED] = $date->format(MYSQL_DATETIME);

        if ($id) {
            $success = $this->update($id, $data);
            if($success) {
                return $id;
            } else {
                return false;
            }
        } else {
            $data[static::CREATED] = $date->format(MYSQL_DATETIME);
            return $this->insert($data);
        }
    }

    /**
     * Retrieve a single record based on an arbitary key=>value WHERE clause
     *
     * @param array $where           An arbitary WHERE clause determining which records to return
     * @param bool  $include_deleted Include deleted records in the result set (defaults to false)
     *
     * @return object|null The object matching the specified query(if any)
    */
    public function get_by($where, $include_deleted = false)
    {
        $this->run_before_callbacks('get');

        if (!$include_deleted) {
            $this->db->where([static::DELETED => null]);
        }

        $row = $this->db->where($where)
                        ->get($this->table)
                        ->row();

        return $this->run_after_callbacks('get', [$row]);
    }
}
```

models/User.php

```php
<?php

namespace App\Models;

use App\Core\Model;

/**
 * Abstraction of user related business logic in the application
 */
class User extends Model
{
    /**
     * {@inheritdoc}
     */
    public function save($data, $id = null)
    {
        if (array_key_exists('password', $data) && strlen($data['password']) > 0) {
            $data['password'] = self::password_hash($data['password']);
        } else {
            unset($data['password']);
        }

        return parent::save($data, $id);
    }

    /**
     * Ensure a username and password match up to the values in the database
     *
     * @param string $username The username of the user to verify the password for
     * @param string $password The unhashed password supplied by the user in plain text
     *
     * @return bool True for valid, false for invalid
    */
    public function password_verify($username, $password)
    {
        $user = $this->get_by(['username' => $username]);
        if ($user) {
            return password_verify($password, $user->password);
        } else {
            return false;
        }
    }

    /**
     * Hash a password using PHP's `password_hash()` function
     *
     * @param string $password The password to hash
     *
     * @return string A password hash containing the salt and hashing algorithm used
    */
    private static function password_hash($password)
    {
        return password_hash($password, PASSWORD_DEFAULT);
    }
}
```

Test:
```php
<?php
use PHPUnit\Framework\TestCase;

use App\Models\User as UserModel;

class User extends TestCase
{
    use \Mockery\Adapter\Phpunit\MockeryPHPUnitIntegration;

    /**
     * @var UserModel
     */
    private $user_model;

    /**
     * @var Object
     */
    private $test_user;

    public function setUp()
    {
        $this->user_model = Mockery::mock(UserModel::class)->makePartial();

        $this->test_user = (object)[
            'username' => 'mrtesting',
            'password' => password_hash('abc', PASSWORD_DEFAULT)
        ];
    }

    /**
     * @test
    */
    public function is_granted_login_with_a_valid_username_and_password()
    {
        $this->user_model->shouldReceive('get_by')
                         ->with(['username' => 'mrtesting'])
                         ->andReturn($this->test_user);

        $valid = $this->user_model->password_verify('mrtesting', 'abc');

        $this->assertTrue($valid);
    }

    /**
     * @test
    */
    public function password_can_be_changed()
    {
        $model = Mockery::mock(UserModel::class . '[update]')->shouldAllowMockingProtectedMethods();

        $model->shouldReceive('update')
              ->once()
              ->with(6, [
                    'password'         => 'hello',
                    UserModel::UPDATED => date(MYSQL_DATETIME)
                ]);

        $model->save(['password' => password_hash('hello', PASSWORD_DEFAULT)], 6);
    }

    public function tearDown()
    {
        Mockery::close();
    }

}
```

**Comment by robertbasic** [comment]
You are expecting the `update` method to receive the following arguments:

 - an integer of value 6
 - an array with with two elements, where the one with the key of `password` should have the value `hello`

but when actually called, the array element with the `password` key has the value of `$2y$10$3zhUkgt9OlneZAhgku62TuDIKgT1XQFfiaUeb4HTrI3xHuoHybCai`. And in that case, Mockery can't find a correct match and throws the exception.

That's one thing that breaks, the other that **will** break sooner or later is the `UserModel::UPDATED` datetime value. It can and it will happen that between setting up the expectation and when the method gets called a second passes and again Mockery will receive different arguments than what's expected and throw an exception.

**Comment by robertmain** [comment]
Yeah, I was wondering about that, but I couldn't figure out if the method signature was somehow wrong, if I'd set the mock up incorrectly (wrong number of arguments, wrong object etc.) or if (as you pointed out) the data I was setting it to expect was wrong. Maybe the exception could be changed to reflect that more specifically?

I updated my mock to look like this:
```php
public function password_can_be_changed()
{
    $model = Mockery::mock(UserModel::class . '[update]')->shouldAllowMockingProtectedMethods();

    $model->shouldReceive('update')
            ->with(6, [
                'password'         => '$2y$10$8LKJXYTAghwmY1q8PfjQQOP22jA6cBC6/WLx6Y.d0fWeylG.XWhjC' //hardcoded for debugging the test...
            ]);

    $model->save(['password' => 'hello'], 6);
}
```
Which produces the following exception:  

`Mockery\Exception\NoMatchingExpectationException: No matching handler found for Mockery_1_App_Models_User::update(6, ['password' => '$2y$10$PzN9EkxEx5pU/MxngFg0suRZyhYc2SzYUFgFlCeawKXFa2eTjh/zS', 'updated_at' => '2018-02-26 13:14:40']). Either the method was unexpected or its arguments matched no expected argument list for this method`

Running the same test again produces: 
`Mockery\Exception\NoMatchingExpectationException: No matching handler found for Mockery_1_App_Models_User::update(6, ['password' => '$2y$10$foT09XbL47qzZ0Iq5VT.ceVZO4b/k/e1s.AOYWUt/I/uTcXdaxsdm', 'updated_at' => '2018-02-26 13:15:51']). Either the method was unexpected or its arguments matched no expected argument list for this method`

Clearly the password hash is being seeded by a PRNG(which I should've seen coming a mile off in retrospect), so I obviously can't make assertions about it's value. That being the case, how would you recommend testing this?

--- 


>the other that will break sooner or later is the UserModel::UPDATED datetime value. It can and it will happen that between setting up the expectation and when the method gets called a second passes and again Mockery will receive different arguments than what's expected and throw an exception.

That's an excellent point and is something I've been wondering about for a while. The only reason that's in there is because the `save()` method adds it automatically. Personally, I couldn't care less about the value of that field for the purposes of this test case. Is it possible to just omit it and have the mock focus solely on the value of `password`?

**Comment by robertbasic** [comment]
> Maybe the exception could be changed to reflect that more specifically?

Not sure how can it get any more clear, although I'm biased, looking at that error message for half a decade now :D 

 > "No matching handler found for X::Y(Z). Either the method was unexpected or its arguments matched no expected argument list for this method"

So either the method call was not expected, or was expected, but with different arguments than what it actually received.

> That being the case, how would you recommend testing this?

You have multiple solutions, depending on what do you want to assert.

That only the `password` key is set?

```` php
$model->shouldReceive('update')
            ->with(6, \Mockery::hasKey('password'));
````

or something more complicated, with `Mockery::on`:

```` php
$model->shouldReceive('update')
            ->with(6, \Mockery::on(function ($argument) {
                        if (!isset($argument['password'])) {
                                    return false;
                        }
                        return substr($argument['password'], 7) == '$2y$10$';
            }));
````

Read more about the [on matcher in the docs](http://docs.mockery.io/en/latest/cookbook/mockery_on.html)

There could probably be other ways to do it as well, but these 2 come to mind without spending much time on looking at the actual code.

**Comment by robertmain** [comment]
>Not sure how can it get any more clear, 

Well, basically what threw me for a loop was that it just basically said "Something is wrong". I wasn't sure if that "something" was me spelling the method name wrong, or maybe I'm mocking the wrong method, maybe it has the wrong number of args etc. I never thought to look at the actual value of the password field to notice that it was changing each time. The fact that I was somewhat sleep deprived yesterday probably didn't help either 🙃

Something like the whole phpunit "expected" vs "actual" thing to show that what it expected to receive and what it actually received data aren't the same
![image](https://user-images.githubusercontent.com/537684/36680707-896e85c6-1ae4-11e8-8acd-3ac883845983.png)


>although I'm biased, looking at that error message for half a decade now :D

Yeah, I totally get that, haha. Some of my apps or libraries have error messages that seem clear to me and then I show it someone else and they ask a question that makes me think for a second. Maybe it could throw [something like this](https://github.com/crazycodr/standard-exceptions#data-exceptions) for errors related to the test data? Dunno. Just a thought :) 

>That only the password key is set?

Basically, I'm trying to assert that when someone sends in `password = foo` that (or rather a secure hash of that) gets saved to the database. I think I have more of a gameplan figured out now after a night's sleep :)

**Comment by robertbasic** [comment]
> Something like the whole phpunit "expected" vs "actual" thing

Ehh... there were some attempts at that I think. I should probably look into that again, but, yeah, time and all that.

> Maybe it could throw something like this for errors related to the test data?

I need to take a closer look at how the whole "unmatched" method and see what can be improved, but again, time :D 

Let me know if you have any more questions, and can this issue be closed.

Thanks!

**Comment by robertmain** [comment]
Yeah, I totally get the whole time thing. Sometimes there seem not to be enough hours in the day.

Otherwise, I'm all set. Thanks for your help :)

**Comment by robertmain** [comment]
@robertbasic Sorry to harp on again - but I've run into a similar(ish) issue with expectations. Given the following test:

```php
public function password_remains_unchanged_if_not_suppled()
{
    $this->user_model->shouldNotreceive('password_hash');

    $this->user_model->shouldReceive('update')
                        ->once()
                        ->with(6, Mockery::on(function($data) {
                        return !array_key_exists('password', $data);
                        }));

    $this->user_model->save(['firstname' => 'bob'], 6);
}
```

If I deliberately break that test and pass the `password` field in, which will in turn cause `UserModel::password_hash()` to be called. The exception I get says _"Call to a member function where() on null"_ (where() being a method that would be called elsewhere in my call stack), rather than say that that `UserModel::password_hash()` was called and it shouldn't have been.

**Comment by robertbasic** [comment]
@robertmain how do you know it gets called? Looks to me that the entire `if` check in `User::save` makes sure that it doesn't get called.

**Comment by robertbasic** [comment]
``` php
    public function save($data, $id = null)
    {
        if (array_key_exists('password', $data) && strlen($data['password']) > 0) {
            $data['password'] = self::password_hash($data['password']);
        } else {
            unset($data['password']);
        }

        return parent::save($data, $id);
    }
```

You're calling the save always, regardless is password supplied or not.

**Comment by robertbasic** [comment]
Oh, and you're creating a generated partial mock where you said you want to mock only the `update` method. So Mockery is mocking only that.

**Comment by robertmain** [comment]
Sure, but that seems to be exactly what [isn't] happening in certain circumstances..

**Comment by robertbasic** [comment]
OK, you've lost me now to be honest. Let's start over. What are you trying to do and what's happening? Please provide only the relevant code and test parts, as well as full error output.

**Comment by robertmain** [comment]
...i should probably open another issue maybe?
